File: 0220-contains-duplicate-iii/0220-contains-duplicate-iii.py
class Solution:
    def containsNearbyAlmostDuplicate(self, nums: List[int], indexDiff: int, valueDiff: int) -> bool:
        bucket = {}
        for i, num in enumerate(nums):
            idx = num // (valueDiff + 1)
            if idx in bucket:
                return True
            if idx - 1 in bucket and abs(num - bucket[idx - 1]) <= valueDiff:
                return True
            if idx + 1 in bucket and abs(num - bucket[idx + 1]) <= valueDiff:
                return True
            bucket[idx] = num
            if i >= indexDiff:
                del bucket[nums[i - indexDiff] // (valueDiff + 1)]
        return False

File: 0220-contains-duplicate-iii/0220-contains-duplicate-iii.js
/**
 * @param {number[]} nums
 * @param {number} indexDiff
 * @param {number} valueDiff
 * @return {boolean}
 */
var containsNearbyAlmostDuplicate = function(nums, indexDiff, valueDiff) {
  const bucket = {};
  const w = valueDiff + 1;
  
  for (let [i, num] of nums.entries()) {
    const idx = Math.floor(num / w);
    if (idx in bucket) return true;
    if (idx - 1 in bucket && Math.abs(num - bucket[idx - 1]) < w)
        return true
    if (idx + 1 in bucket && Math.abs(num - bucket[idx + 1]) < w)
        return true
    
    bucket[idx] = num;
    if (i >= indexDiff) {
      const oldNum = nums[i - indexDiff];
      delete bucket[Math.floor(oldNum / w)];
    }
  }
  
  return false;
};